### ISSUE TITLE
🚀 Feature: Gated Scaffolder Workflows

### ISSUE THREAD

**Issue by taras** [backstage/backstage#16622]
### 🔖 Feature description

We need a way to created Scaffolder workflows that includes actions that may span days. 

There are a bunch of use cases that need this, for example:

1. Wait for pull request to be merged - when a scaffolder workflow creates a pull request, we might want to wait for the pull request to be merged before continuing. Since merging pull requests requires human intervention, it may take days if not weeks for that action to complete.
2. Wait for another user to provide additional input - when creating workflows that require approval from someone within the organization, we might need to wait for the user to press a button before the workflow continues.
3. Wait for a resource to be created - when creating services, you might want to wait until a resource like a database is created before proceeding.

There is no way to do this currently.

### 🎤 Context

At the last Adoption SIG, I demoed using the [Embedded Scaffolder Workflow](https://www.youtube.com/watch?v=vdpm4m7uMgg) to onboard an entity to use TechDocs by creating a pull request with everything necessary to onboard the entity. This demo is missing a few details that are required to create a complete user experience. 

We need to handle the following states,
1. The workflow ran and created a pull request that needs to be merged. We need to show to the user that the PR is waiting.
2. After the PR is merged, but before the YAML document is ingested, we need to show to the user that onboarding is pending ingestion.

Handing these states consistently is best handled by the framework because there are many ways to mess this up. The Gated Scaffolder Workflows will make handling these use cases very easy. 

### ✌️ Possible Implementation

We need to create an RFC to flash out how this will work.

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

**Comment by bforbis** [comment]
Curious how this might relate to the Prefect workflow in the Box demo during this community session?
https://youtu.be/apCDT3_DmFk?t=669

**Comment by stephenmuss** [comment]
Is there any movement on this at all? Such as an RFC. I'd be interested to see this.

**Comment by efenner-cambia** [comment]
Any movement on this? I would also like to see this and it looks like the other issue that was linked above got auto-closed

**Comment by taras** [comment]
We're about to ship a project using the Embedded Scaffolder plugin. I'm hoping we'll have the opportunity to work on this next because it'll help to provide status on previous executions of scaffolder.

**Comment by sharadpattanshetti** [comment]
> We're about to ship a project using the Embedded Scaffolder plugin. I'm hoping we'll have the opportunity to work on this next because it'll help to provide status on previous executions of scaffolder.

Any example available on how this can be used in the templates?

**Comment by deeparavirdc** [comment]
+1 Is there some documentation or example on how to enable this in backstage and use it with templates ?

**Comment by npalm** [comment]
Looking forward to gated support!

**Comment by efenner-cambia** [comment]
@taras Any update?

**Comment by Alok650** [comment]
Following! This would be a very useful feature.

**Comment by benjdlambert** [comment]
Related #20934

**Comment by knksmith57** [comment]
Bump to keep this alive 🙏

**Comment by fvmiranda** [comment]
Keep this alive!

**Comment by fvmiranda** [comment]
@taras do you have any updates ?

**Comment by taras** [comment]
@fvmiranda Not really; the client that needed this has other priorities now. We have a few ideas on how to build it, but we need bandwidth to do it.

**Comment by Alok650** [comment]
Hi guys, 
Has anyone created such a feature? I have multiple usecases for gating mechanism i.e. an approval (maker-checker) based workflow mechanism.
@benjdlambert is it part of the scaffolder roadmap?

**Comment by JamesAtIntegratnIO** [comment]
Still looking for this feature. 

Bot reset comment

**Comment by benjdlambert** [comment]
@Alok650 the recoverable tasks work which is still experimental is going to help with this, but it's not currently prioritized to work on yet. We still need to work out how we're going to affectively pause and resume tasks, which might be powered with something like the `events-backend`, but still needs planning a little bit I think.

**Comment by siegenthalerroger** [comment]
Still looking for this feature.

Bot reset comment

**Comment by nia-potato** [comment]
same here.

**Comment by jgrumboe** [comment]
Bot reset

**Comment by benjdlambert** [comment]
Hey :wave: update on this is that I think that we're pretty much ready to explore how we would implement this.

We now have workspace serialization, and events support in the `scaffolder`, so wondering if anyone would like to startup a BEP for how we should implement this?

Thinking that to start we should rely on the `events-backend` for an http ingress to be able to resume existing jobs, and some way to mark jobs as pending and paused.

**Comment by abhipsnl** [comment]
> Hey 👋 update on this is that I think that we're pretty much ready to explore how we would implement this.
> 
> We now have workspace serialization, and events support in the `scaffolder`, so wondering if anyone would like to startup a BEP for how we should implement this?
> 
> Thinking that to start we should rely on the `events-backend` for an http ingress to be able to resume existing jobs, and some way to mark jobs as pending and paused.

Did anyone initiate this?

**Comment by benjdlambert** [comment]
@abhipsnl nope not as yet, feel free to start one if you would like to contribute? :pray:

**Comment by brandon-adams-wwt** [comment]
beep beep boop boop bot reset 🤖

**Comment by andrevcf** [comment]
Reset

**Comment by twoodardgit** [comment]
Does anyone have any updates on this feature request?

**Comment by benjdlambert** [comment]
Think that https://github.com/backstage/backstage/issues/16622#issuecomment-2511207271 is still the latest of this issue. Looking for someone to help drive this forward still :pray:

**Comment by jwillaz** [comment]
This is a pretty coveted feature. @benjdlambert - what is the initial spark needed to get this rolling? You previously mentioned workspace serialization and events support in the scaffolder. Could you say more about how those would work with this feature? I would love to help add thoughts where I can, but the blocker has been understanding where these underlying core changes come into play.

**Comment by andrevcf** [comment]
Reset

**Comment by andrevcf** [comment]
any updates on this?

**Comment by twoodardgit** [comment]
I know this probably isn't the easiest to implement but I think this is probably the single most coveted scaffolder workflow enhancement being requested currently for the application. Is there any potential updates on when the primary maintainers might look to add this or prioritize it?

**Comment by rnowacoski** [comment]
bump

**Comment by andrevcf** [comment]
Bump

**Comment by andrevcf** [comment]
Bump

**Comment by spiffaz** [comment]
Hey @benjdlambert - picking up your Dec 2024 note asking someone to start a BEP for this. I've written a full draft and wanted to socialise it on the issue before pushing it forward, since the owners would need to be the scaffolder maintainers.

It takes the shape you sketched: a durable `waiting` (pending/paused) task state, and a resume path through the events backend to bring parked jobs back. Built on the workspace-serialisation and checkpoint work from BEP-0004 rather than a new execution model.

The load-bearing decisions:

- `ctx.suspend(...)` parks the task and releases the worker (a control-signal unwind), so a wait never holds a `PQueue` slot - the direct answer to the "it holds up workers" concern behind the debug:wait cap (#24846).
- Resume replays the suspended action's handler from the top, with `suspend` acting as a checkpoint that returns the resume payload, reusing the BEP-0004 replay-and-skip machinery instead of adding a coroutine model.
- Task consumption is untouched - pickup stays `claim()`/poll. The only event-driven surface is the inbound resume signal, so it deliberately doesn't reopen the consumption rewrite from #13799.
- Core ships only the primitive; approvals, wait-for-merge and wait-for-resource are plugins on top. No orchestration semantics, so it's not the direction #23967 was closed against.

Honest about scope: it depends on BEP-0004's recovery/serialisation, which is still experimental, so there's a real sequencing decision written up as an open question. There's also an interim approval plugin that works today with zero core changes, for anyone who needs this before the primitive lands.

Draft PR: https://github.com/backstage/backstage/pull/34966

Would one of the scaffolder maintainers be up for co-sponsoring / owning it? Happy to take it to the SIG once it's back on, but wanted a maintainer attached first. Feedback on the shape very welcome.
